### DISCUSSION TITLE
122

### DISCUSSION THREAD

**Discussion by jojoldu** [jojoldu/blog-comments#714] [Comments]
http://jojoldu.tistory.com/122

**Comment by jojoldu** [comment]
작성자: 강현구

`List<EnumModel>` 이 아닌 `List<EnumValue>` 쓴 이유를 알 수 있을까요??

**Comment by jojoldu** [comment]
작성자: 창천향로 창천향로

앗 방문 감사드립니다!
EnumValue를 별도로 구현해서 사용한 이유는 List<EnumModel>을 Controller에서 View로 전달하게 되면 Enum의 key만 전달되기 때문입니다. 
아래 코드를 Controller에 추가해서 테스트 해보시면 될것 같습니다.

```java
    @GetMapping("/models")
    public List<EnumModel> getModel() {
        return Arrays
                .stream(EnumContract.CommissionType.class.getEnumConstants())
                .collect(Collectors.toList());
    }
```

다시한번 댓글 감사드립니다^^
혹시나 제가 알고 있는 지식이 틀리다면 댓글 부탁드리겠습니다!

**Comment by jojoldu** [comment]
작성자: Kingbbode

정말 좋은 글 감사합니다!

점진적으로 좀 더 관리하기 편하고 사용하기 편한 코드 발전해나가는 방식으로 설명해주셔서 이해가 잘 되었습니다!

조금 아쉬운 것이 EnumMapper를 빈으로 등록할 때 Enum을 하나씩 put하는 것이 조금 번거로워 보입니다!
자바 리플렉션 활용한 패키지 스캔 방식을 사용하여 자동으로 등록하는 것은 어떨까요?

**Comment by jojoldu** [comment]
작성자: 창천향로 창천향로

앗 좋은의견 감사드립니다^^
말씀주신대로 리플렉션을 이용해서 패키지 단위 혹은 어노테이션 단위로 스캐닝 하는 방법을 생각했었지만, 이럴 경우 해당 Enum들이 어떻게 관리되는지 한눈에 파악하기가 어렵다는 생각에 Bean 메소드에 직접 추가하는 방식을 선택하게 되었습니다.

예를들어 현재는 enum이 한 패키지에 다 있지만 프로젝트 설계에 따라 여러 domain class 내부에 존재할 수도 있습니다. 
이럴 경우 어느 enum은 mapper에서 관리 대상에서 제외시키고 싶고, 어느 enum은 mapper에서 관리시키고 싶다고 한다면 리플렉션으로는 한눈에 파악하기가 힘들다고 생각합니다.
대신 AppConfig에서 직접 put하게 되면 현재 어떤 enum들이 enumMapper에 등록되어있고, get용 key값이 어떻게 되는지 한눈에 볼 수 있다고 생각합니다.

개인적인 생각이지만 추상화 레벨을 올릴수록 코드는 간결해지는 반면, 파악하기는 힘들어진다고 생각합니다.

이런 이유로 위처럼 구현하게 되었습니다
개인적인 생각이라 kingbbode님의 말씀대로 패키지 혹은 어노테이션 단위로 관리도 괜찮다고 생각합니다^^

좋은의견 다시한번 감사드립니다!

**Comment by jojoldu** [comment]
작성자: Kingbbode

 존경스럽습니다! 감사합니다!

**Comment by jojoldu** [comment]
작성자: 창천향로 창천향로

앗 덜덜... 그렇지는 않습니다 ㅠㅠ
kingbbode 님 글 보면서 많이 공부하고 있습니다.
제가 더감사합니다^^

**Comment by jojoldu** [comment]
작성자: youngjae

글을 읽는데 토비의 스프링을 읽는 느낌이네요ㅎㅎ

enum은 항상 따로 빼서 사용했는데 공통 메소드는 인터페이스로 두고 같은 성격의 요소들을 도메인 내에 모아서 쓰는게 더 깔끔한 구조가 되는것 같습니다.

한 가지 아쉬웠던 점이 코드가 흑백이라는것이...ㅋㅋ

항상 감사합니다~

**Comment by jojoldu** [comment]
작성자: 창천향로 창천향로

으억 ㅋㅋ... 토비님의 스프링과는 정말 1000배 차이나는 퀄리티인지라 ㅠㅠ... 말씀만이라도 감사합니다.

위 코드는 github 코드 스타일이 적용되서 어쩔수 없는것 같습니다 ㅠ 

항상 방문해주셔서 감사합니다!!

**Comment by jojoldu** [comment]
작성자: 준티

정말 많은 도움이 됐습니다 ^^

**Comment by ShipJH** [comment]
잘봤습니다.

다만 MONEY란 데이터로  저장되었을때 뽑아낼때 금액으로 보여주고싶은데.. 
(예를들면) 리스트성 데이터에 
NO | TYPE | CONTENT | .....
1      MONEY
2     ....
3

이렇게 들어가있을때 

조회해서 가져올때 MONEY는 금액으로 값을 바꿔주는 일련의 List를 재정의하는 함수 처리 방법말고는 없겠죠?

**Comment by toongri** [comment]
선생님 감사합니다! 이 글 덕에 코드가 훨씬 직관적으로 변했습니다!
정말 감사드립니다 코드보고 너무 뿌듯하네요ㅠㅠ
